### ISSUE TITLE
User permissions data in store is referenced globally

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#5772]
## Bug Description

The `capabilities` data in the `core/user` store is one of the few bits of data in the store which is resolved from a global variable on page load. This is fine, but the way that it is received preserves the reference to the object defined on the global, such that the `permissions` object is the same as the `capabilities` object in the data store and changes to one are reflected in both. This is not a vulnerability, as all the same capability checks are enforced on the server side (where all the more sensitive information and behavior lives), but more of a flaw in the implementation that violates the purpose of Redux, which is that state can only be modified by actions.

## Steps to reproduce

On a Site Kit screen, this condition evaluates to `true`
```js
_googlesitekitUserData.permissions === googlesitekit.data.select('core/user').getCapabilities()
```

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* User `capabilities` state should not be mutable by reference

## Implementation Brief

- Remove the following code:
https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/includes/Core/Permissions/Permissions.php#L237-L243
- Add the permissions route to the preloaded paths list using the `googlesitekit_apifetch_preload_paths` filter: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/includes/Core/Permissions/Permissions.php#L705-L718
- Rename the `fetchRefreshCapabilitiesStore` fetch store to be `getCapabilities` instead: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/assets/js/googlesitekit/datastore/user/permissions.js#L40-L47
- Rework the `getCapabilities` resolver to use the renamed `getCapabilities` fetch store to get permissions: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/assets/js/googlesitekit/datastore/user/permissions.js#L165-L179
- Update the `refreshCapabilities` action to just refetch capabilities using the renamed `getCapabilities` fetch store: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/assets/js/googlesitekit/datastore/user/permissions.js#L111-L126

### Test Coverage

- Fix any failing tests.
- Update the existing tests for the above change.

## QA Brief

* This issue works on refactoring the way how user permissions are sourced from the server to the client. This doesn't have any visual change.
* Verify that calling `_googlesitekitUserData.permissions` in the browser console on a Site Kit screen returns `undefined`.
* Verify that Site Kit works as expected overall. Little more attention should be given to the following views:
    * Dashboard Sharing settings modal.
    * Module Recovery for disconnected recoverable modules.
* Provide 'Editor' user role access to all of the modules in Dashboard Sharing settings modal.
* Log in as the editor user.
* When visiting the view-only dashboard verify that the widgets appear appropriately and there's no error in the console, as described [here](https://github.com/google/site-kit-wp/issues/5772#issuecomment-1266890649).

## Changelog entry

* N/A

**Comment by eugene-manuilov** [comment]
> * In the `refreshCapabilities` action, remove the code which sets the `global._googlesitekitUserData.permissions`:

@hussain-t why do we need to remove that code? For me it looks correct and we shouldn't change it. The only thing that I think we need to do there is to decouple it from the object passed to the `receiveCapabilities` action.

**Comment by aaemnnosttv** [comment]
> @hussain-t why do we need to remove that code? For me it looks correct and we shouldn't change it. The only thing that I think we need to do there is to decouple it from the object passed to the `receiveCapabilities` action.

@eugene-manuilov while I don't think we necessarily need to _remove_ that code as part of this issue, I do feel we probably shouldn't be updating our globals at runtime. Since we need to fetch capabilities now for refreshing, maybe we should update this selector to use a preloaded response and then refreshing would simply invoke the fetch again rather than using a global at all. What do you think?

**Comment by hussain-t** [comment]
@aaemnnosttv, should I remove that part from the IB to keep the IB specific to the issue so we can tackle updating the global variable separately?

**Comment by aaemnnosttv** [comment]
@hussain-t SGTM 👍

**Comment by hussain-t** [comment]
@aaemnnosttv, this will not be fixed if we don't address updating the global variable.

Let's say we execute this condition after addressing the fix only made in the `getCapabilities` resolver, and then we execute the same condition after dispatching the `receiveCapabilities` action; it still evaluates to `true`, as you can see in the screenshot:

![Screenshot 2022-09-07 at 10 25 19 AM](https://user-images.githubusercontent.com/24408261/188792874-72414e04-8ed0-4e62-b076-b41c98967c6f.png)


The problem is that the code incorrectly resets the entire global object of `_googlesitekitUserData` (not only the `permissions`) with the `response`. You can execute `_googlesitekitUserData` before and after dispatching the `receiveCapabilities` action:

![Screenshot 2022-09-07 at 10 28 15 AM](https://user-images.githubusercontent.com/24408261/188793110-677fcfd1-2753-427c-b4ee-c09be7b09496.png)


IMO, we should fix it at least by setting only the `permissions` object as part of this issue:

```js
global._googlesitekitUserData.permissions = {
	...response,
};
```

WDYT?

**Comment by eugene-manuilov** [comment]
@hussain-t I think what Evan says is a more appropriate way of resolving the issue. Instead of dealing with global variables, we should use our API preload functionality to preload "refreshed capabilities" and then use it as a starting point for user capabilities. Here is a list of steps that we need to do:

1. Remove https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/includes/Core/Permissions/Permissions.php#L237-L243
1. Add the permissions route to the preloaded paths list using the `googlesitekit_apifetch_preload_paths` filter: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/includes/Core/Permissions/Permissions.php#L705-L718
1. Rename the `fetchRefreshCapabilitiesStore` fetch store to be `getCapabilities` instead: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/assets/js/googlesitekit/datastore/user/permissions.js#L40-L47
1. Rework the `getCapabilities` resolver to use the renamed `getCapabilities` fetch store to get permissions: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/assets/js/googlesitekit/datastore/user/permissions.js#L165-L179
1. Update the `refreshCapabilities` action to just refetch capabilities using the renamed `getCapabilities` fetch store: https://github.com/google/site-kit-wp/blob/412f9126c6bbb6736faa33107d1a3346232e21b4/assets/js/googlesitekit/datastore/user/permissions.js#L111-L126

If this approach is okay to you, I think we can use it as IB then 😄. cc @aaemnnosttv

**Comment by eugene-manuilov** [comment]
@aaemnnosttv if said in the previous comment looks good to you, then do we need to update AC?

**Comment by techanvil** [comment]
@eugene-manuilov, while I see nothing wrong with your general point about moving from these inline JS snippets to the preloaded routes system, when it comes to the particular issue we are trying to solve here, this might end up just kicking the can a bit down the road. It turns out we are also sharing references between our globally visible preloaded data and Redux. For example:

![image](https://user-images.githubusercontent.com/18395600/188872426-af897491-d8f8-4702-8509-17a5822f9161.png)

cc @aaemnnosttv

**Comment by eugene-manuilov** [comment]
I agree with you, @techanvil. Do you mind creating a new ticket for preloaded data? We can fix all preloaded data issues all together.

**Comment by aaemnnosttv** [comment]
> If this approach is okay to you, I think we can use it as IB then

Yeah, I was also thinking we should rename to `getCapabilities` as the refresh is really just invoking a fetch and updating state like a normal fetch store 👍  My only concern about this change is that it will change capabilities to no longer be available immediately on load, so we may need to do something extra to account for that.

@techanvil that's a good point about the preloaded data too. Turns out we have a lot of global references kicking around! 😆 

I think we can keep this issue to reworking around the capabilities specifically as @eugene-manuilov outlined above and tackle things more holistically in a separate issue. For now I think we can simply copy the `permissions` when we receive them in the fetch store's `reducerCallback`. It could be this is all we do in other places too but it's worth exploring  other possibilities in another issue.

**Comment by hussain-t** [comment]
Thanks, @aaemnnosttv @eugene-manuilov, and @techanvil. I have updated the IB to copy the `permissions` within the `reducerCallback` and update the `getCapabilities` resolver accordingly. LMK if this looks good. I think we can simply copy the `permissions` as part of this issue and tackle things more holistically in #5798.

@aaemnnosttv WDYT?

**Comment by aaemnnosttv** [comment]
@hussain-t this is still in progress, correct?

**Comment by hussain-t** [comment]
@aaemnnosttv, I have updated the IB to reflect @eugene-manuilov's comment. Thanks!

**Comment by techanvil** [comment]
Thanks, @eugene-manuilov and @aaemnnosttv, I have created a separate issue to tackle the preloaded data: https://github.com/google/site-kit-wp/issues/5832

**Comment by eugene-manuilov** [comment]
Thanks, @hussain-t. IB ✔️

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@nfmohit I am unsure if this is related to this ticket but I have an observation.  

While testing another ticket,  I enabled Dashboard sharing and wanted to make sure that the widgets were appearing as expected on the main and entity dashboard in view only.  I gave 'Editor' user role access to all of the modules, and logged in as the editor user.   When I visit the view only dashboard no widgets appear and an error occurs in the console.

```
googlesitekit-vendor-f5043c9bdc48cc7d9190.js:1          GET https://5859.us10.instawp.xyz/index.php/wp-json/google-site-kit/v1/core/user/data/permissions?_locale=user 403
googlesitekit-api-8384c6dab354771044f7.js:1 Google Site Kit API Error method:GET datapoint:permissions type:core identifier:user error:"Sorry, you are not allowed to do that."
```
When I logged in as an administrator and visited the view only dashboard, the widgets loaded as expected. 

![image](https://user-images.githubusercontent.com/73545194/193788438-878636c1-b252-4091-8b11-cb8120f5a0ad.png)

Please could you check if this issue is related to the changes here, if not, I will create a new ticket.

**Comment by nfmohit** [comment]
Thank you for spotting this, @wpdarren! I've added [a new PR](https://github.com/google/site-kit-wp/pull/5966) to address this.

**Comment by hussain-t** [comment]
@nfmohit, could you update the QAB to verify this [issue](https://github.com/google/site-kit-wp/issues/5772#issuecomment-1266890649) is fixed?

**Comment by nfmohit** [comment]
Updated QAB, thanks @hussain-t!

**Comment by aaemnnosttv** [comment]
Reopening this one due to a regression flagged on https://github.com/google/site-kit-wp/issues/5435#issuecomment-1270530080 where permissions are not available right away as before which can cause some odd flickers in some cases where we didn't have them before due to the immediate availability via the global. This was my initial concern with this change as mentioned in https://github.com/google/site-kit-wp/issues/5772#issuecomment-1242405051. We can still do it, but we need to make sure permissions are always available right away, at least for now.

**Comment by aaemnnosttv** [comment]
Created a PR to fix this for the release, will open another issue to address this in a better way in the future. This is still an improvement over what we had I think, but the PR preserves the timing/availability of the state as we had before when it was just received from a global.

**Comment by makiost** [comment]
The [follow-up PR](https://github.com/google/site-kit-wp/pull/5977) fixes the issue flagged in https://github.com/google/site-kit-wp/issues/5435#issuecomment-1270530080. Tested it locally, and the availability ensures that we don't encounter any wrong texts from showing up while the `googlesitekit_manage_module_sharing_options` option is undefined.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- Calling` _googlesitekitUserData.permissions` in the browser console on a Site Kit screen returns false.  The QAB says it should read undefined but after chatting with Hussain, this is correct. 
- I set up Site Kit for all the modules and they connected without any issues. Permissions appeared as expected. Also tested the Dashboard Sharing settings modal and Module Recovery for disconnected recoverable modules.
-  For Dashboard Sharing as an editor, the view-only dashboard verify that the widgets appear appropriately and there's no error in the console, as described https://github.com/google/site-kit-wp/issues/5772#issuecomment-1266890649.
- For the flicker issue reported [here](https://github.com/google/site-kit-wp/issues/5435#issuecomment-1270530080) by Mohit, this is now resolved on the main branch. 

**Note:** we will pick up any regressions running release QA checks. From an overview check, everything looks as expected. 

![image](https://user-images.githubusercontent.com/73545194/194325353-a5d6e330-7525-4fee-a32b-15272dca87bc.png)
![image](https://user-images.githubusercontent.com/73545194/194325402-64fab002-4a6f-42d5-b550-ea07bbf8d860.png)

**Comment by nfmohit** [comment]
Thank you @aaemnnosttv for spotting the issue and for the quick fix!
